### PR TITLE
drivers: pinctrl: renesas: fix ipsr generation for Spider boards

### PR DESCRIPTION
**Pull request by xakep-amatop**
Avoid unexpected memory access in cases where the IPSR has an odd register index. In the scenario where an odd register index is present in IPSR, the LSB of the register index is utilized as MSB of the bank number. Observe how we pack reg/bank:
    'IPSR(((reg) << 4U) | (bank), shift, func)' (macro IPnSR)
and how it is read from the device tree source:
    '(RCAR_IPSR(node_id) >> 10U) & 0x1FU' (macro RCAR_PIN_FUNC).

 Finally, this bank is used to obtain the required IPSR offset:
    'PFC_RCAR_IPSR + rcar_func->bank * sizeof(uint32_t)'
in the 'pfc_rcar_set_ipsr' function. For example, if we have the 1 as a reg index and the 3-rd bank, the resulting offset would be 19, which is beyond the IPSR range.

Align the IPSR comment with the definition of the 'rcar_pin_func' structure.

Add missed definitions of IP[0-3]SR[67] macros.

Note: we can omit the usage of the register index entirely since this information is obtained from the pin number inside the 'pfc_rcar_get_reg_index' function.

### CODE REVIEW

File: include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h [hunk 1]
```diff
@@ -45,7 +47,7 @@
  * Each base address has 4 IPSR banks.
  */
 #define IPnSR(bank, reg, shift, func) \
-	IPSR(((reg) << 4U) | (bank), shift, func)
+	IPSR(((reg) << 5U) | (bank), shift, func)
```
Comment (lorc) [inline include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:50]: I am not sure that I understand how it worked before. Are you saying that all definitions with non-zero `reg` were broken?

File: include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h [hunk 2]
```diff
@@ -71,6 +73,14 @@
 #define IP1SR5(shift, func) IPnSR(1, 5, shift, func)
 #define IP2SR5(shift, func) IPnSR(2, 5, shift, func)
 #define IP3SR5(shift, func) IPnSR(3, 5, shift, func)
+#define IP0SR6(shift, func) IPnSR(0, 6, shift, func)
```
Comment (lorc) [inline include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:76]: I believe that this should go into separate patch

**Review by lorc — CHANGES_REQUESTED** [review]

File: include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h (hunk 1 above)
Comment (xakep-amatop) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:50]: There were only broken definitions with odd register indices, such as 1, 3, 5... As far as I can see, they weren't used before. In the Spider ARM32 dts, I've found pins with IPxSR0 (UART), so this is probably the reason why we don't face this issue.

Comment (lorc) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:50]: I still not sure that I got it. As I understand, this is not just about LSB, but about all the bits in `reg`.

E.g. previously `IPnSR(1, 5, shift, func)` would result `bank` value to be `0x15`, but with your change it will be `0x25`, which is different.   And of course mentioned code

```
PFC_RCAR_IPSR + rcar_func->bank * sizeof(uint32_t)
```

will generate completely different offset which of course is beyond the range 0-18.

Comment (xakep-amatop) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:50]: if we examine the behavior with an even register index with the previous code, it would be the same as with my patch. For example, considering IPnSR(2, 5, shift, func), in the previous version, we would have 0x25, and in the new one, 0x45. After masking with 0x1F, we will obtain the same result in the `bank` field.

File: include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h (hunk 2 above)
Comment (xakep-amatop) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:76]: done

File: include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h [hunk 3]
```diff
@@ -17,8 +17,10 @@
  * @param func the 4 bits encoded alternate function.
  *
  * Function code    [ 0 : 3 ]
- * Function shift   [ 4 : 9 ]
- * IPSR bank        [ 10 : 13 ]
+ * Function shift   [ 4 : 8 ]
+ * Empty            [ 9 ]
+ * IPSR bank        [ 10 : 14 ]
+ * Register index   [ 15 : 17 ] (Spider only)
```
[the remark this answers is not in the document]
Comment (xakep-amatop) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h]: done

File: include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h (hunk 1 above)
Comment (lorc) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:50]: Okay, I missed the masking with 0x1F. But now I am confused even more: why do we need `reg` field in the first place? It is basically ignored by the pinctrl driver.

Anyways, this is the completely different question. Thank you for the clarification.

Comment (xakep-amatop) [reply include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:50]: yes, I mentioned this in a commit message:
```
Note: we can omit the usage of the register index entirely since this information is obtained from the pin number inside the 'pfc_rcar_get_reg_index' function.
```
We can remove it entirely, but it requires a lot of changes in many files. Therefore, I decided that it will be easier to fix a shift of the register index.

**Review by lorc — APPROVED** [review]
Well, I believe that S4 code will use this additional information in the future, so it is better to leave this field.

**Comment by xakep-amatop** [comment]
> Well, I believe that S4 code will use this additional information in the future, so it is better to leave this field.

thank you for a review

**Comment by aaillet** [comment]
Hi,

Thanks for the PR, I confirmed that both SCIF0 and SCIF3 pins are still working fine (located on IP0SR0).
Were you able to confirm everything is alright using pins from odd registers ?

**Comment by xakep-amatop** [comment]
> Hi,
> 
> Thanks for the PR, I confirmed that both SCIF0 and SCIF3 pins are still working fine (located on IP0SR0). Were you able to confirm everything is alright using pins from odd registers ?

A logical mistake was detected during the review of the pfc driver code. Currently, we don't have devices that use IPxSR[1357] registers. At the moment, only two devices are in use, which require pin control: hscif0 (I'm working on adding it right now) and MMC. However, there are no issues with either of them.

I can share with you a small patch with compile time checks --- build fails in case when a bank outside the defined range
pls, let me now if you need it

**Review by aaillet — APPROVED** [review]
@xakep-amatop shared is result with me on discord, using a build time check, we can confirm that this patch is needed to stay in available range when using pins from odd registers

LGTM
